### PR TITLE
Lms 71,74

### PR DESCRIPTION
**Pull request by DemchyshynV**

### CODE REVIEW

File: src/database/models/user.schema.ts [hunk 1]
```diff
@@ -25,7 +25,8 @@ UserSchema = new Schema({
     },
     email: {
         type: String,
-        required: true
+        required: true,
+        unique: true
```
Comment (DemchyshynV) [inline src/database/models/user.schema.ts]: I think it should be here or not?

File: src/services/user/user.service.ts [hunk 2]
```diff
@@ -1,17 +1,49 @@
 import { model } from 'mongoose';
+
+import { DatabaseTablesEnum } from '../../constants/enums';
 import { User, UserSchema, UserType } from '../../database';
 import { IUser } from '../../interfaces';
 
 class UserService {
-    createUser(userValue: IUser): Promise<any> {
+    async createUser(userValue: IUser): Promise<any> {
```
Comment (VictorKmin) [inline src/services/user/user.service.ts]: use async just if you have await in the method body

File: src/services/user/user.service.ts [hunk 3]
```diff
@@ -1,17 +1,49 @@
...
     }
-    getUserByParams(params: Partial<IUser>) {
-        const UserModel = model<UserType>('User', UserSchema);
 
+    async getUserByParams(params: Partial<IUser>): Promise<any> {
+        const UserModel = model<UserType>(DatabaseTablesEnum.USER_COLLECTION_NAME, UserSchema);
         return UserModel.findOne(params);
     }
+
+    async changeStatus(user_id: string, status: number): Promise<any> {
```
Comment (VictorKmin) [inline src/services/user/user.service.ts]: the same. Check all

File: src/routes/user/user.router.ts [hunk 4]
```diff
@@ -1,11 +1,25 @@
...
-router.post('/:user_id/block', userController.blockUser);
+router.get('/', userController.getAll);
+router.get('/allAdmins', userController.getAllAdmins);
+router.get('/allTeachers', userController.getAllTeachers);
+router.get('/allStudents', userController.getAllStudents);
+router.get('/getInfo', userController.getUserInfoByToken);
+router.get('/:user_id', userController.getByID);
+
+router.post('/',checkIsEmailPresent, userController.createUser);
+
```
Comment (VictorKmin) [inline src/routes/user/user.router.ts]: router.use(isUserPresent)

File: src/routes/user/user.router.ts [hunk 5]
```diff
@@ -1,11 +1,25 @@
...
 
-router.post('/', userController.createUser);
-router.get('/getInfo', userController.getInfo);
-router.post('/:user_id/block', userController.blockUser);
+router.get('/', userController.getAll);
+router.get('/allAdmins', userController.getAllAdmins);
+router.get('/allTeachers', userController.getAllTeachers);
+router.get('/allStudents', userController.getAllStudents);
+router.get('/getInfo', userController.getUserInfoByToken);
+router.get('/:user_id', userController.getByID);
```
Comment (VictorKmin) [inline src/routes/user/user.router.ts]: move to line 17. Need to check is user present

File: src/routes/user/user.router.ts [hunk 6]
```diff
@@ -1,11 +1,25 @@
...
 import { userController } from '../../controllers';
+import {checkIsEmailPresent} from "../../middleware/auth";
 
 const router = Router();
 
-router.post('/', userController.createUser);
-router.get('/getInfo', userController.getInfo);
-router.post('/:user_id/block', userController.blockUser);
+router.get('/', userController.getAll);
+router.get('/allAdmins', userController.getAllAdmins);
```
Comment (VictorKmin) [inline src/routes/user/user.router.ts]: router.get('/users', userController.getAllUsers);
And filter in query params.
/users?role=1
/users?role=2

File: src/routes/user/user.router.ts [hunk 7]
```diff
@@ -1,11 +1,25 @@
...
+router.get('/allTeachers', userController.getAllTeachers);
```
Comment (VictorKmin) [inline src/routes/user/user.router.ts]: rm

File: src/routes/user/user.router.ts [hunk 8]
```diff
@@ -1,11 +1,25 @@
...
+router.get('/allStudents', userController.getAllStudents);
```
Comment (VictorKmin) [inline src/routes/user/user.router.ts]: rm

File: src/routes/user/user.router.ts [hunk 9]
```diff
@@ -1,11 +1,25 @@
...
+router.get('/getInfo', userController.getUserInfoByToken);
```
Comment (VictorKmin) [inline src/routes/user/user.router.ts]: rm

File: src/routes/user/user.router.ts [hunk 10]
```diff
@@ -1,11 +1,25 @@
 import { Router } from 'express';
 
 import { userController } from '../../controllers';
+import {checkIsEmailPresent} from "../../middleware/auth";
```
Comment (VictorKmin) [inline src/routes/user/user.router.ts]: import {checkIsEmailPresent} from "../../middleware";

File: src/controllers/user/user.controller.ts [hunk 11]
```diff
@@ -1,56 +1,200 @@
...
+            }
+
+            res.json(result);
+
+        } catch (e) {
+            next(e);
+        }
+    }
+
+    async getAllAdmins(req: IRequestExtended, res: Response, next: NextFunction) {
```
Comment (VictorKmin) [inline src/controllers/user/user.controller.ts]: merge into 1 method 
/users and filter by role
/users?role=1
/users?status=1
etc,

File: src/controllers/user/user.controller.ts [hunk 12]
```diff
@@ -1,56 +1,200 @@
...
-        } catch (e) {
-            next(e);
-        }
-    }
-    async getInfo(req: IRequestExtended, res: Response, next: NextFunction) {
-        try {
-            const  { _id, name, group_id, photo_path, role_id, status_id, surname, email }  = req.user as IUser;
-            const user = {
+            const {_id, email, name, surname, role_id, status_id, photo_path, groups_id} = req.user as IUser;
+            const result: IUserSubjectModel = {
```
Comment (MikolaLvivUA) [inline src/controllers/user/user.controller.ts]: maybe better user or userSubject but not result. That little strange label.

File: src/controllers/user/user.controller.ts [hunk 13]
```diff
@@ -1,56 +1,200 @@
...
-                group_id,
-                photo_path,
                 role_id,
-                status_id
+                status_id,
+                photo_path,
+                groups_id
             };
-            res.json(user);
+            res.json(result);
```
Comment (MikolaLvivUA) [inline src/controllers/user/user.controller.ts]: please change on data: user or userSubject. We have that style everywhere.

File: src/controllers/user/user.controller.ts [hunk 14]
```diff
@@ -1,56 +1,200 @@
...
+        } catch (e) {
+            next(e);
+        }
+    }
+
+    async blockUser(req: IRequestExtended, res: Response, next: NextFunction) {
+        try {
+            const {user_id} = req.params;
+            await userService.changeStatus(user_id, UserStatusEnum.BLOCKED);
+            res.end();
```
Comment (MikolaLvivUA) [inline src/controllers/user/user.controller.ts]: I think better send some massage like res.json(`user ${user_id} has been blocked`)

File: src/controllers/user/user.controller.ts [hunk 15]
```diff
@@ -1,56 +1,200 @@
...
+        } catch (e) {
+            next(e);
+        }
+    }
+
+    async unBlockUser(req: IRequestExtended, res: Response, next: NextFunction) {
+        try {
+            const {user_id} = req.params;
+            await userService.changeStatus(user_id, UserStatusEnum.ACTIVE);
+            res.end();
```
Comment (MikolaLvivUA) [inline src/controllers/user/user.controller.ts]: The same

File: src/controllers/user/user.controller.ts [hunk 16]
```diff
@@ -1,56 +1,200 @@
...
+        } catch (e) {
+            next(e);
+        }
+    }
+
+    async changeRoleToAdmin(req: IRequestExtended, res: Response, next: NextFunction) {
+        try {
+            const {user_id} = req.params;
+            await userService.changeRole(user_id, UserRoleEnum.ADMIN);
+            res.end();
```
Comment (MikolaLvivUA) [inline src/controllers/user/user.controller.ts]: the same

File: src/controllers/user/user.controller.ts [hunk 17]
```diff
@@ -1,56 +1,200 @@
...
+        } catch (e) {
+            next(e);
+        }
+    }
+
+    async changeRoleToTeacher(req: IRequestExtended, res: Response, next: NextFunction) {
+        try {
+            const {user_id} = req.params;
+            await userService.changeRole(user_id, UserRoleEnum.TEACHER);
+            res.end();
```
Comment (MikolaLvivUA) [inline src/controllers/user/user.controller.ts]: the same

File: src/controllers/user/user.controller.ts [hunk 18]
```diff
@@ -1,56 +1,200 @@
...
+        } catch (e) {
+            next(e);
+        }
+    }
+
+    async changeRoleToStudent(req: IRequestExtended, res: Response, next: NextFunction) {
+        try {
+            const {user_id} = req.params;
+            await userService.changeRole(user_id, UserRoleEnum.STUDENT);
+            res.end();
```
Comment (MikolaLvivUA) [inline src/controllers/user/user.controller.ts]: the same

File: src/controllers/user/user.controller.ts [hunk 19]
```diff
@@ -1,56 +1,200 @@
...
+        }
+    }
+
+    async delete(req: IRequestExtended, res: Response, next: NextFunction) {
+        try {
+            const {user_id} = req.params;
+            await userService.delete(user_id);
+            res.end();
+        } catch (e) {
+            next(e);
```
Comment (MikolaLvivUA) [inline src/controllers/user/user.controller.ts]: the same look All controollers

File: src/controllers/user/user.controller.ts [hunk 20]
```diff
@@ -1,56 +1,200 @@
...
+        }
+    }
+
+    async getAll(req: IRequestExtended, res: Response, next: NextFunction) {
+        try {
+            const {_id} = req.user as IUser;
+            const result = await userService.getAll(_id) as [IUser];
```
Comment (MikolaLvivUA) [inline src/controllers/user/user.controller.ts]: The same change on const users

File: src/controllers/user/user.controller.ts [hunk 21]
```diff
@@ -1,56 +1,200 @@
...
+
+            if (!result) {
+                return next(new ErrorHandler(
+                    ResponseStatusCodesEnum.NOT_FOUND,
+                    errors.NOT_FOUND_USER_NOT_PRESENT.message,
+                    errors.NOT_FOUND_USER_NOT_PRESENT.code
+                ));
+            }
+
+            res.json(result);
```
Comment (MikolaLvivUA) [inline src/controllers/user/user.controller.ts]: data: users

File: src/controllers/user/user.controller.ts [hunk 22]
```diff
@@ -1,56 +1,200 @@
...
+        } catch (e) {
+            next(e);
+        }
+    }
+
+    async getByID(req: IRequestExtended, res: Response, next: NextFunction) {
+        try {
+            const {user_id} = req.params;
+            const result = await userService.getByID(user_id) as IUser;
```
Comment (MikolaLvivUA) [inline src/controllers/user/user.controller.ts]: const user check all controllers

File: src/controllers/user/user.controller.ts [hunk 23]
```diff
@@ -1,56 +1,200 @@
...
+
+            if (!result) {
+                return next(new ErrorHandler(
+                    ResponseStatusCodesEnum.NOT_FOUND,
+                    errors.NOT_FOUND_USER_NOT_PRESENT.message,
+                    errors.NOT_FOUND_USER_NOT_PRESENT.code
+                ));
+            }
+
+            res.json(result);
```
Comment (MikolaLvivUA) [inline src/controllers/user/user.controller.ts]: data: user Check all controllers

File: src/services/user/user.service.ts [hunk 24]
```diff
@@ -1,17 +1,49 @@
...
+        const UserModel = model<UserType>(DatabaseTablesEnum.USER_COLLECTION_NAME, UserSchema);
+        return UserModel.findByIdAndUpdate(user_id, { status_id: status });
+    }
+
+    async changeRole(user_id: string, role: number): Promise<any> {
+        const UserModel = model<UserType>(DatabaseTablesEnum.USER_COLLECTION_NAME, UserSchema);
+        return UserModel.findByIdAndUpdate(user_id, { role_id: role });
+    }
+
+    async delete(user_id: string): Promise<any> {
```
Comment (MikolaLvivUA) [inline src/services/user/user.service.ts]: in a method like this better use Promise<void>. please check all method

File: src/routes/user/user.router.ts [hunk 25]
```diff
@@ -1,25 +1,21 @@
 import { Router } from 'express';
 
 import { userController } from '../../controllers';
-import {checkIsEmailPresent} from "../../middleware/auth";
+import { checkIsEmailPresent } from '../../middleware/auth';
+import { isUserPresent } from '../../middleware/user';
```
Comment (VictorKmin) [inline src/routes/user/user.router.ts]: import { checkIsEmailPresent, isUserPresent } from '../../middleware';

**Pull request merged**